packages/sdk/app-graph/src/graph-builder.ts

```diff
@@ -2,50 +2,236 @@
 // Copyright 2023 DXOS.org
 //
 
-import { EventSubscriptions, type UnsubscribeCallback } from '@dxos/async';
+import { type Signal, effect, signal } from '@preact/signals-core';
+// import { yieldOrContinue } from 'main-thread-scheduling';
+
+import { EventSubscriptions } from '@dxos/async';
+import { invariant } from '@dxos/invariant';
 
 import { Graph } from './graph';
+import { type EdgeDirection, type NodeArg, type Node } from './node';
+
+export type ConnectorExtension = (params: {
+  node: Node;
+  direction: EdgeDirection;
+  type?: string;
+}) => NodeArg<any>[] | undefined;
+export type HydratorExtension = (params: { id: string; type: string }) => NodeArg<any> | undefined;
+
+export type ExtensionType = 'connector' | 'hydrator';
+export type BuilderExtension =
+  | {
+      type: 'connector';
+      extension: ConnectorExtension;
+    }
+  | {
+      type: 'hydrator';
+      extension: HydratorExtension;
+    };
+
+export const connector = (extension: ConnectorExtension): BuilderExtension => ({ type: 'connector', extension });
+export const hydrator = (extension: HydratorExtension): BuilderExtension => ({ type: 'hydrator', extension });
+
+export type GraphBuilderTraverseOptions = {
+  node: Node;
+  direction?: EdgeDirection;
+  visitor: (node: Node, path: string[]) => void;
+};
+
+class BuilderInternal {
```

Comment:
i don't understand why this needs to be static and it feels dangerous.

Reply:
strictly speaking it doesn't need to be static, but the alternative api feels very cumbersome (needing to pass an instance of a builder around everywhere to track the state). The goal here was to end up with a more React-like api where that wasn't necessary.

There are certainly potentials gotchas with this, for instance, everything needs to be using the same version of app-graph otherwise they could end up with different static classes which are misaligned. However, this problem of ensuring versions are aligned is also a problem that we already deal with in other packages that we use so that didn't feel like a blocker.

All that said, if there was a way to have this _not_ be static or module level state or something like but still have a nice clean builder api then I'd love that, I just haven't figured out a way to accomplish that to this point.